src/tyr/isochrone_serializer.cc

```diff
@@ -192,73 +247,143 @@ std::string serializeGeoTIFF(Api& request, const std::shared_ptr<const GriddedDa
+  for (const auto metric_idx : std::views::iota(0U, metrics.size())) {
+    if (!metrics[metric_idx]) {
+      continue;
+    }
+    auto current_plane = planes[metric_idx];
+    current_plane.reserve(static_cast<size_t>(ext_x) * static_cast<size_t>(ext_y));
+    std::vector<uint16_t> data(static_cast<size_t>(ext_x) * static_cast<size_t>(ext_y));
 
     // seconds or 10 meter steps
-    float scale_factor = metric_idx == 0 ? 60 : 100;
-    for (int32_t i = 0; i < ext_y; ++i) {
-      for (int32_t j = 0; j < ext_x; ++j) {
+    const float scale_factor = (metric_idx == 0) ? 60.f : 100.f;
+    for (const auto i : std::views::iota(0, ext_y)) {
+      for (const auto j : std::views::iota(0, ext_x)) {
         auto tileid = isogrid->TileId(j + box[0], i + box[1]);
+        // flip Y so first row is top
         data[(ext_y - 1 - i) * ext_x + j] =
             static_cast<uint16_t>(isogrid->DataAt(tileid, metric_idx) * scale_factor);
       }
     }
+    planes[metric_idx] = std::move(data);
+  }
 
-    auto band = geotiff_dataset->GetRasterBand(nbands == 2 ? (metric_idx + 1) : 1);
-    band->SetNoDataValue(std::numeric_limits<uint16_t>::max());
-    band->SetDescription(metric_idx == 0 ? "Time (seconds)" : "Distance (10m)");
-
-    CPLErr err = band->RasterIO(GF_Write, 0, 0, ext_x, ext_y, data, ext_x, ext_y, GDT_UInt16, 0, 0);
-
-    delete[] data;
+  // in-memory TIFF
+  MemTIFFStream mem;
+  TIFF* tif =
+      XTIFFClientOpen("mem", "w", mem.handle(), MemTIFFStream::read, MemTIFFStream::write,
+                      MemTIFFStream::seek, MemTIFFStream::close, MemTIFFStream::size,
+                      static_cast<TIFFMapFileProc>(nullptr), static_cast<TIFFUnmapFileProc>(nullptr));
+
+  // Core tags
+  TIFFSetField(tif, TIFFTAG_IMAGEWIDTH, static_cast<uint32_t>(ext_x));
+  TIFFSetField(tif, TIFFTAG_IMAGELENGTH, static_cast<uint32_t>(ext_y));
+  TIFFSetField(tif, TIFFTAG_BITSPERSAMPLE, 16);
+  TIFFSetField(tif, TIFFTAG_SAMPLESPERPIXEL, valid_bands);
+  TIFFSetField(tif, TIFFTAG_PLANARCONFIG, PLANARCONFIG_SEPARATE); // one plane per band
+  TIFFSetField(tif, TIFFTAG_PHOTOMETRIC,
+               PHOTOMETRIC_MINISBLACK); // required, hints at B/W display
+  TIFFSetField(tif, TIFFTAG_COMPRESSION, COMPRESSION_PACKBITS);
+  TIFFSetField(tif, TIFFTAG_SAMPLEFORMAT, SAMPLEFORMAT_UINT);
+  TIFFSetField(tif, TIFFTAG_ORIENTATION, ORIENTATION_TOPLEFT);
+
+  // we let libtiff decide how to optimize binning, more interesting for compression
+  uint32_t rowsperstrip = TIFFDefaultStripSize(tif, 0);
+  if (rowsperstrip == 0 || rowsperstrip > static_cast<uint32_t>(ext_y))
+    rowsperstrip = static_cast<uint32_t>(ext_y);
+  TIFFSetField(tif, TIFFTAG_ROWSPERSTRIP, rowsperstrip);
+
+  // it seems geotiff as we know them have some gdal custom tags
+  // TODO: write NODATA values:
+  //   - it _should_ be possible with the metadata xml below, e.g.
+  //     <Item name="NODATA_VALUES">{} {}</Item> </GDALMetadata>)", isogrid->MaxValue(0),
```

Comment:
I tried this, didn't work for QGIS, probably better to initialize the grid with max(uint16_t)